File: project-{topic}-complete/src/backend/services/product_service.cpp
```cpp
#include "product_service.h"

ProductService::ProductService(DBManager& db_manager)
    : db_manager_(db_manager), product_cache_(), logger_(spdlog::get("ecommerce_logger")) {
    if (!logger_) {
        logger_ = spdlog::stdout_color_mt("product_service_logger");
    }
}

Product ProductService::create_product(const Product& product) {
    if (product.name.empty() || product.price < 0 || product.stock < 0) {
        throw BadRequestException("Invalid product data: Name, price, and stock are required and must be valid.");
    }

    // Check if product name already exists
    try {
        auto products = db_manager_.get_all_products(product.name, 1, 0);
        for(const auto& p : products) {
            if (p.name == product.name) {
                throw ConflictException("Product with this name already exists.");
            }
        }
    } catch (const DBGenericException& e) {
        // If the search query fails, we can proceed, but log it.
        logger_->warn("Failed to check for existing product name during creation: {}", e.what());
    }

    Product created_product = db_manager_.create_product(product);
    product_cache_.invalidate(created_product.id); // Invalidate cache for all products if it was cached
    product_cache_.put(created_product.id, created_product); // Cache the new product
    logger_->info("Product created: ID={}, Name={}", created_product.id, created_product.name);
    return created_product;
}

Product ProductService::get_product_by_id(int id) {
    // Try to get from cache first
    std::optional<Product> cached_product = product_cache_.get(id);
    if (cached_product) {
        return cached_product.value();
    }

    std::optional<Product> product_opt = db_manager_.get_product_by_id(id);
    if (!product_opt) {
        throw NotFoundException("Product with ID " + std::to_string(id) + " not found.");
    }

    Product product = product_opt.value();
    product_cache_.put(id, product); // Cache the retrieved product
    logger_->info("Product retrieved: ID={}, Name={}", product.id, product.name);
    return product;
}

std::vector<Product> ProductService::get_all_products(const std::string& search_term, int limit, int offset) {
    // For search, limit, offset, caching entire result sets is complex.
    // We'll rely on DB for these queries directly for now.
    // A more advanced cache would involve query caching with invalidation strategies.
    logger_->info("Retrieving all products with search='{}', limit={}, offset={}", search_term, limit, offset);
    return db_manager_.get_all_products(search_term, limit, offset);
}

Product ProductService::update_product(const Product& product) {
    // Basic validation
    if (product.id <= 0) {
        throw BadRequestException("Invalid product ID for update.");
    }
    if (!product.name.empty() && product.name.length() < 3) {
        throw BadRequestException("Product name must be at least 3 characters long.");
    }
    if (product.price < 0) {
        throw BadRequestException("Product price cannot be negative.");
    }
    if (product.stock < 0) {
        throw BadRequestException("Product stock cannot be negative.");
    }

    Product updated_product = db_manager_.update_product(product);
    product_cache_.invalidate(updated_product.id); // Invalidate cache for this product
    product_cache_.put(updated_product.id, updated_product); // Update cache with new product details
    logger_->info("Product updated: ID={}, Name={}", updated_product.id, updated_product.name);
    return updated_product;
}

void ProductService::delete_product(int id) {
    if (id <= 0) {
        throw BadRequestException("Invalid product ID for delete.");
    }
    db_manager_.delete_product(id);
    product_cache_.invalidate(id); // Invalidate cache for this product
    logger_->info("Product deleted: ID={}", id);
}
```

File: project-{topic}-complete/backend/src/services/productService.js
```javascript
const { Product, Category, Review } = require('../config/db');
const { AppError } = require('../utils/appError');
const logger = require('../utils/logger');
const { setCache, getCache, deleteCache } = require('../utils/cache');

const PRODUCTS_CACHE_KEY = 'all_products';

/**
 * Service for Product related business logic.
 */
class ProductService {
    /**
     * Fetches all products with optional filtering, sorting, and pagination.
     * Caches the result.
     * @param {Object} queryOptions - Options for filtering, sorting, pagination.
     * @returns {Array<Object>} List of products.
     */
    static async getAllProducts(queryOptions = {}) {
        const {
            page = 1,
            limit = 10,
            sort = 'createdAt',
            order = 'DESC',
            categoryId,
            search
        } = queryOptions;

        const offset = (page - 1) * limit;

        const where = {};
        if (categoryId) {
            where.categoryId = categoryId;
        }
        if (search) {
            where.name = {
                [Product.sequelize.Op.iLike]: `%${search}%`
            };
        }

        // Generate a unique cache key for these query options
        const cacheKey = `${PRODUCTS_CACHE_KEY}_${JSON.stringify(queryOptions)}`;
        const cachedProducts = getCache(cacheKey);
        if (cachedProducts) {
            logger.debug(`Serving products from cache: ${cacheKey}`);
            return cachedProducts;
        }

        const products = await Product.findAndCountAll({
            where,
            include: [{
                model: Category,
                as: 'category',
                attributes: ['id', 'name']
            }],
            order: [[sort, order]],
            limit: parseInt(limit, 10),
            offset: parseInt(offset, 10)
        });

        // Store in cache
        setCache(cacheKey, products);
        logger.debug(`Products fetched from DB and cached: ${cacheKey}`);
        return products;
    }

    /**
     * Fetches a single product by ID.
     * Caches the result.
     * @param {string} productId - ID of the product.
     * @returns {Object} Product object.
     */
    static async getProductById(productId) {
        const cacheKey = `product_${productId}`;
        const cachedProduct = getCache(cacheKey);
        if (cachedProduct) {
            logger.debug(`Serving product from cache: ${cacheKey}`);
            return cachedProduct;
        }

        const product = await Product.findByPk(productId, {
            include: [
                { model: Category, as: 'category', attributes: ['id', 'name'] },
                { model: Review, as: 'reviews', attributes: ['id', 'rating', 'comment', 'userId', 'createdAt'] }
            ]
        });

        if (!product) {
            throw new AppError('Product not found', 404);
        }

        setCache(cacheKey, product);
        logger.debug(`Product fetched from DB and cached: ${cacheKey}`);
        return product;
    }

    /**
     * Creates a new product (Admin only).
     * @param {Object} productData - Product data.
     * @returns {Object} Created product object.
     */
    static async createProduct(productData) {
        const { categoryId } = productData;
        const category = await Category.findByPk(categoryId);
        if (!category) {
            throw new AppError('Category not found.', 400);
        }
        const newProduct = await Product.create(productData);
        deleteCache(PRODUCTS_CACHE_KEY); // Invalidate all product listings cache
        logger.info(`Product created: ${newProduct.name}`);
        return newProduct;
    }

    /**
     * Updates an existing product (Admin only).
     * @param {string} productId - ID of the product to update.
     * @param {Object} updateData - Data to update.
     * @returns {Object} Updated product object.
     */
    static async updateProduct(productId, updateData) {
        const product = await Product.findByPk(productId);
        if (!product) {
            throw new AppError('Product not found', 404);
        }

        if (updateData.categoryId) {
            const category = await Category.findByPk(updateData.categoryId);
            if (!category) {
                throw new AppError('Category not found.', 400);
            }
        }

        await product.update(updateData);
        deleteCache(`product_${productId}`); // Invalidate specific product cache
        deleteCache(PRODUCTS_CACHE_KEY);     // Invalidate all product listings cache
        logger.info(`Product updated: ${product.name} (ID: ${productId})`);
        return product;
    }

    /**
     * Deletes a product (Admin only).
     * @param {string} productId - ID of the product to delete.
     */
    static async deleteProduct(productId) {
        const product = await Product.findByPk(productId);
        if (!product) {
            throw new AppError('Product not found', 404);
        }
        await product.destroy();
        deleteCache(`product_${productId}`);
        deleteCache(PRODUCTS_CACHE_KEY);
        logger.info(`Product deleted: ${productId}`);
    }

    // --- Category Services (could be in a separate CategoryService) ---
    /**
     * Fetches all categories.
     * @returns {Array<Object>} List of categories.
     */
    static async getAllCategories() {
        // Categories typically don't change often, can cache
        const cacheKey = 'all_categories';
        const cachedCategories = getCache(cacheKey);
        if (cachedCategories) {
            return cachedCategories;
        }

        const categories = await Category.findAll();
        setCache(cacheKey, categories);
        return categories;
    }

    /**
     * Creates a new category (Admin only).
     * @param {Object} categoryData - Category data.
     * @returns {Object} Created category object.
     */
    static async createCategory(categoryData) {
        const newCategory = await Category.create(categoryData);
        deleteCache('all_categories'); // Invalidate category cache
        logger.info(`Category created: ${newCategory.name}`);
        return newCategory;
    }
}

module.exports = ProductService;
```